File: server/services/drawn-geometry.ts
/**
 * Drawn-geometry authoring service (US-001).
 *
 * Pure helpers that turn a geometry drawn on the map (a GeoJSON Polygon or
 * LineString) into a reviewable contribution. Nothing here touches fs / express
 * / the source-of-truth TSVs — the drawn shape lands in the *contribution queue*
 * (via `ContributionService.submit`) with `provenance source = 'user-drawn'`, so
 * a human reviewer promotes it into `data/source/lexicons/*.tsv` later (US-009).
 *
 * The geometry is stored in the same shape the geometry TSVs use: GeoJSON with
 * standard `[lng, lat]` coordinate order (see `data/source/lexicons/civilization-boundaries
 * .tsv`, `migration-routes.tsv`). `serializeGeometry` produces exactly the JSON
 * string that would sit in a TSV `geometry`/`waypoints` cell.
 */

import type { ContributionEntityType, ContributionSource, Contribution } from "./contribution-service";

// ============================================================================
// Types
// ============================================================================

/** GeoJSON position — `[longitude, latitude]` (optional elevation ignored). */
export type Position = [number, number] | [number, number, number];

export interface DrawnPolygon {
  type: "Polygon";
  /** One or more linear rings; each ring is a closed list of positions. */
  coordinates: Position[][];
}

export interface DrawnLineString {
  type: "LineString";
  coordinates: Position[];
}

export type DrawnGeometry = DrawnPolygon | DrawnLineString;

/**
 * Where a drawn shape is destined once a reviewer promotes it. Polygon targets
 * describe an area; line targets describe a path. This maps 1:1 onto the
 * geometry-bearing TSVs and onto `ContributionEntityType`.
 */
export type DrawnGeometryTarget =
  | "boundary" // civilization-boundaries.tsv (Polygon)
  | "language-range" // language-range-polygons.tsv (Polygon)
  | "trade-route" // trade-routes.tsv (LineString)
  | "migration-route"; // migration-routes.tsv (LineString)

const POLYGON_TARGETS: DrawnGeometryTarget[] = ["boundary", "language-range"];
const LINE_TARGETS: DrawnGeometryTarget[] = ["trade-route", "migration-route"];

export const DRAWN_GEOMETRY_TARGETS: DrawnGeometryTarget[] = [
  ...POLYGON_TARGETS,
  ...LINE_TARGETS,
];

export interface DrawnGeometryInput {
  geometry: DrawnGeometry;
  target: DrawnGeometryTarget;
  /** Human-readable label for the drawn feature. */
  name: string;
  /** The existing entity this geometry is associated with (required). */
  associatedEntityId: string;
  /** Start year (negative = BCE). Required — a geometry must be time-anchored. */
  timePeriodStart: number;
  /** End year; if omitted the feature is treated as a point-in-time at start. */
  timePeriodEnd?: number | null;
  timePeriodLabel?: string;
  /** 1–100. Defaults to 60 for hand-drawn geometry. */
  confidence?: number;
  sources?: ContributionSource[];
  description?: string;
  notes?: string;
  contributorName?: string;
  contributorEmail?: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

/** Marks the provenance of a drawn contribution's `entityData.source`. */
export const DRAWN_PROVENANCE = "user-drawn" as const;

// ============================================================================
// Geometry validation
// ============================================================================

function isFiniteNumber(n: unknown): n is number {
  return typeof n === "number" && Number.isFinite(n);
}

/** A single `[lng, lat]` position with in-bounds coordinates. */
function isValidPosition(pos: unknown): pos is Position {
  return (
    Array.isArray(pos) &&
    pos.length >= 2 &&
    isFiniteNumber(pos[0]) &&
    isFiniteNumber(pos[1]) &&
    pos[0] >= -180 &&
    pos[0] <= 180 &&
    pos[1] >= -90 &&
    pos[1] <= 90
  );
}

function positionsEqual(a: Position, b: Position): boolean {
  return a[0] === b[0] && a[1] === b[1];
}

/**
 * Validate the structural GeoJSON of a drawn geometry. Rules:
 * - Polygon: ≥1 ring, each ring ≥4 positions and closed (first === last).
 * - LineString: ≥2 positions.
 * - Every position must be a finite `[lng, lat]` within world bounds.
 */
export function validateGeometry(geometry: unknown): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (!geometry || typeof geometry !== "object") {
    return { valid: false, errors: ["geometry is required and must be an object"], warnings };
  }

  const geom = geometry as { type?: unknown; coordinates?: unknown };

  if (geom.type === "Polygon") {
    const rings = geom.coordinates;
    if (!Array.isArray(rings) || rings.length === 0) {
      errors.push("Polygon geometry must have at least one linear ring");
    } else {
      rings.forEach((ring, ri) => {
        if (!Array.isArray(ring) || ring.length < 4) {
          errors.push(`Polygon ring ${ri} must have at least 4 positions (a closed triangle)`);
          return;
        }
        ring.forEach((pos, pi) => {
          if (!isValidPosition(pos)) {
            errors.push(`Polygon ring ${ri} position ${pi} is not a valid [lng, lat] within world bounds`);
          }
        });
        const first = ring[0];
        const last = ring[ring.length - 1];
        if (isValidPosition(first) && isValidPosition(last) && !positionsEqual(first, last)) {
          errors.push(`Polygon ring ${ri} is not closed (first and last positions must match)`);
        }
      });
    }
  } else if (geom.type === "LineString") {
    const coords = geom.coordinates;
    if (!Array.isArray(coords) || coords.length < 2) {
      errors.push("LineString geometry must have at least 2 positions");
    } else {
      coords.forEach((pos, pi) => {
        if (!isValidPosition(pos)) {
          errors.push(`LineString position ${pi} is not a valid [lng, lat] within world bounds`);
        }
      });
    }
  } else {
    errors.push(`Unsupported geometry type: ${String(geom.type)} (expected Polygon or LineString)`);
  }

  return { valid: errors.length === 0, errors, warnings };
}

/**
 * Canonical serialization of a drawn geometry — the exact JSON string that would
 * live in a geometry-bearing TSV cell. Coordinates keep `[lng, lat]` order.
 */
export function serializeGeometry(geometry: DrawnGeometry): string {
  return JSON.stringify(geometry);
}

// ============================================================================
// Input validation
// ============================================================================

function expectedGeometryType(target: DrawnGeometryTarget): "Polygon" | "LineString" {
  return POLYGON_TARGETS.includes(target) ? "Polygon" : "LineString";
}

/**
 * Validate a full drawn-geometry submission: geometry shape, target/geometry
 * agreement, entity association, and a coherent time range.
 */
export function validateDrawnGeometry(input: Partial<DrawnGeometryInput>): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (!input.target || !DRAWN_GEOMETRY_TARGETS.includes(input.target)) {
    errors.push(`target must be one of: ${DRAWN_GEOMETRY_TARGETS.join(", ")}`);
  }

  if (!input.name || typeof input.name !== "string" || !input.name.trim()) {
    errors.push("name is required");
  }

  if (!input.associatedEntityId || typeof input.associatedEntityId !== "string" || !input.associatedEntityId.trim()) {
    errors.push("associatedEntityId is required — a drawn geometry must be associated with an entity");
  }

  // Time range: start required; end optional but, when present, must not precede start.
  if (!isFiniteNumber(input.timePeriodStart)) {
    errors.push("timePeriodStart is required and must be a number (negative = BCE)");
  }
  if (input.timePeriodEnd !== undefined && input.timePeriodEnd !== null) {
    if (!isFiniteNumber(input.timePeriodEnd)) {
      errors.push("timePeriodEnd must be a number or null");
    } else if (isFiniteNumber(input.timePeriodStart) && input.timePeriodEnd < input.timePeriodStart) {
      errors.push("timePeriodEnd must not be earlier than timePeriodStart (inverted range)");
    }
  }

  if (input.confidence !== undefined) {
    if (!isFiniteNumber(input.confidence) || input.confidence < 1 || input.confidence > 100) {
      errors.push("confidence must be a number between 1 and 100");
    }
  } else {
    warnings.push("confidence not specified, defaulting to 60");
  }

  // Geometry structural validation.
  const geomResult = validateGeometry(input.geometry);
  errors.push(...geomResult.errors);

  // Target ↔ geometry type agreement (only if both are individually valid).
  if (
    input.target &&
    DRAWN_GEOMETRY_TARGETS.includes(input.target) &&
    geomResult.valid &&
    input.geometry
  ) {
    const expected = expectedGeometryType(input.target);
    if (input.geometry.type !== expected) {
      errors.push(
        `target '${input.target}' expects a ${expected} geometry but received a ${input.geometry.type}`,
      );
    }
  }

  return { valid: errors.length === 0, errors, warnings };
}

// ============================================================================
// Mapping to a contribution
// ============================================================================

/**
 * Map a validated drawn geometry onto a `Partial<Contribution>` for the review
 * queue. The geometry is stored under `entityData.geometry` (a GeoJSON object,
 * matching `BoundaryDrawingPanel`) alongside its canonical serialized string,
 * and provenance `source: 'user-drawn'` is recorded on the entity data. For
 * `language-range` the associated entity id is mirrored into `languageId` to
 * satisfy the contribution service's required fields.
 */
export function drawnGeometryToContribution(input: DrawnGeometryInput): Partial<Contribution> {
  const entityType = input.target as ContributionEntityType;
  const confidence = input.confidence ?? 60;

  const entityData: Record<string, unknown> = {
    name: input.name,
    geometry: input.geometry,
    geometrySerialized: serializeGeometry(input.geometry),
    drawingMode: input.geometry.type === "Polygon" ? "polygon" : "polyline",
    source: DRAWN_PROVENANCE,
    associatedEntityId: input.associatedEntityId,
    timePeriodStart: input.timePeriodStart,
    timePeriodEnd: input.timePeriodEnd ?? null,
    timePeriodLabel: input.timePeriodLabel,
  };

  if (input.description) entityData.description = input.description;
  if (entityType === "language-range") {
    entityData.languageId = input.associatedEntityId;
  }

  const sources: ContributionSource[] =
    input.sources && input.sources.length > 0
      ? input.sources
      : [{ title: "User-drawn geometry" }];

  return {
    entityType,
    action: "add",
    entityId: input.associatedEntityId,
    entityData,
    sources,
    confidence,
    notes: input.notes,
    contributorName: input.contributorName,
    contributorEmail: input.contributorEmail,
  };
}

File: services/api/src/pinakes/authoring/drawn_geometry.py
"""Drawn-geometry authoring — the port of `server/services/drawn-geometry.ts`.

A GeoJSON Polygon or LineString drawn on the map becomes a queued contribution
with provenance ``entityData.source = "user-drawn"``. The geometry keeps the
corpus's ``[lng, lat]`` order, and :func:`serialize_geometry` produces exactly
the JSON string a `geometry`/`waypoints` TSV cell holds — the promotion target.

The target vocabulary maps 1:1 onto ``ContributionEntityType``
(:data:`~pinakes.contributions.store.REQUIRED_FIELDS`), which is why a
`language-range` mirrors its associated entity into ``entityData.languageId``:
that type requires a `languageId`, and without the mirror a valid drawing would
be rejected by the queue one layer down.
"""

from __future__ import annotations

from typing import Any, Final, NamedTuple

from pinakes.authoring._js import (
    MISSING,
    field,
    has_sources,
    is_finite_number,
    is_present,
    json_text,
    truthy,
)

#: Targets that describe an area (Polygon) and a path (LineString).
POLYGON_TARGETS: Final[tuple[str, ...]] = ("boundary", "language-range")
LINE_TARGETS: Final[tuple[str, ...]] = ("trade-route", "migration-route")

#: Where a drawn shape is destined once a reviewer promotes it.
DRAWN_GEOMETRY_TARGETS: Final[tuple[str, ...]] = POLYGON_TARGETS + LINE_TARGETS

#: Marks the provenance of a drawn contribution's ``entityData.source``.
DRAWN_PROVENANCE: Final = "user-drawn"

#: Confidence hand-drawn geometry gets when the author declares none.
DEFAULT_GEOMETRY_CONFIDENCE: Final = 60


class ValidationResult(NamedTuple):
    valid: bool
    errors: list[str]
    warnings: list[str]


def _is_valid_position(position: Any) -> bool:
    """A single ``[lng, lat]`` with finite, in-world-bounds coordinates."""
    if not isinstance(position, list) or len(position) < 2:
        return False
    lng, lat = position[0], position[1]
    if not is_finite_number(lng) or not is_finite_number(lat):
        return False
    return bool(-180 <= lng <= 180 and -90 <= lat <= 90)


def _positions_equal(a: Any, b: Any) -> bool:
    return bool(a[0] == b[0] and a[1] == b[1])


def validate_geometry(geometry: Any) -> ValidationResult:
    """Structural GeoJSON validation.

    Polygon: ≥1 ring, each ring ≥4 positions and closed. LineString: ≥2
    positions. Every position a finite ``[lng, lat]`` within world bounds.
    """
    errors: list[str] = []
    warnings: list[str] = []

    if not isinstance(geometry, dict):
        return ValidationResult(
            valid=False,
            errors=["geometry is required and must be an object"],
            warnings=warnings,
        )

    kind = geometry.get("type", MISSING)
    coordinates = geometry.get("coordinates", MISSING)

    if kind == "Polygon":
        if not isinstance(coordinates, list) or not coordinates:
            errors.append("Polygon geometry must have at least one linear ring")
        else:
            for ring_index, ring in enumerate(coordinates):
                if not isinstance(ring, list) or len(ring) < 4:
                    errors.append(
                        f"Polygon ring {ring_index} must have at least 4 positions "
                        "(a closed triangle)"
                    )
                    continue
                for position_index, position in enumerate(ring):
                    if not _is_valid_position(position):
                        errors.append(
                            f"Polygon ring {ring_index} position {position_index} "
                            "is not a valid [lng, lat] within world bounds"
                        )
                first, last = ring[0], ring[-1]
                if (
                    _is_valid_position(first)
                    and _is_valid_position(last)
                    and not _positions_equal(first, last)
                ):
                    errors.append(
                        f"Polygon ring {ring_index} is not closed (first and last "
                        "positions must match)"
                    )
    elif kind == "LineString":
        if not isinstance(coordinates, list) or len(coordinates) < 2:
            errors.append("LineString geometry must have at least 2 positions")
        else:
            for position_index, position in enumerate(coordinates):
                if not _is_valid_position(position):
                    errors.append(
                        f"LineString position {position_index} is not a valid "
                        "[lng, lat] within world bounds"
                    )
    else:
        errors.append(
            f"Unsupported geometry type: {_type_text(kind)} "
            "(expected Polygon or LineString)"
        )

    return ValidationResult(valid=not errors, errors=errors, warnings=warnings)


def _type_text(value: Any) -> str:
    """``String(geom.type)`` for the unsupported-type message."""
    if value is MISSING:
        return "undefined"
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def serialize_geometry(geometry: Any) -> str:
    """The exact JSON string a geometry-bearing TSV cell would hold."""
    return json_text(geometry)


def _expected_geometry_type(target: str) -> str:
    return "Polygon" if target in POLYGON_TARGETS else "LineString"


def validate_drawn_geometry(body: Any) -> ValidationResult:
    """Validate a full submission: geometry, target agreement, association, time."""
    errors: list[str] = []
    warnings: list[str] = []

    target = field(body, "target")
    if target not in DRAWN_GEOMETRY_TARGETS:
        errors.append(f"target must be one of: {', '.join(DRAWN_GEOMETRY_TARGETS)}")

    name = field(body, "name")
    if not isinstance(name, str) or not name.strip():
        errors.append("name is required")

    associated = field(body, "associatedEntityId")
    if not isinstance(associated, str) or not associated.strip():
        errors.append(
            "associatedEntityId is required — a drawn geometry must be "
            "associated with an entity"
        )

    start = field(body, "timePeriodStart")
    if not is_finite_number(start):
        errors.append(
            "timePeriodStart is required and must be a number (negative = BCE)"
        )
    end = field(body, "timePeriodEnd")
    if is_present(end):
        if not is_finite_number(end):
            errors.append("timePeriodEnd must be a number or null")
        elif is_finite_number(start) and end < start:
            errors.append(
                "timePeriodEnd must not be earlier than timePeriodStart "
                "(inverted range)"
            )

    confidence = field(body, "confidence")
    if confidence is not MISSING:
        if not is_finite_number(confidence) or confidence < 1 or confidence > 100:
            errors.append("confidence must be a number between 1 and 100")
    else:
        warnings.append(
            f"confidence not specified, defaulting to {DEFAULT_GEOMETRY_CONFIDENCE}"
        )

    geometry = field(body, "geometry")
    geometry_result = validate_geometry(geometry)
    errors.extend(geometry_result.errors)

    # Target ↔ geometry-type agreement, only once both are individually valid —
    # otherwise a malformed shape would be reported twice.
    if (
        target in DRAWN_GEOMETRY_TARGETS
        and geometry_result.valid
        and isinstance(geometry, dict)
    ):
        expected = _expected_geometry_type(target)
        actual = geometry.get("type", MISSING)
        if actual != expected:
            errors.append(
                f"target '{target}' expects a {expected} geometry but received "
                f"a {_type_text(actual)}"
            )

    return ValidationResult(valid=not errors, errors=errors, warnings=warnings)


def drawn_geometry_to_contribution(body: Any) -> dict[str, Any]:
    """Map a validated drawing onto the queue's ``Partial<Contribution>`` shape."""
    entity_type = field(body, "target")
    confidence = field(body, "confidence")
    if confidence is MISSING:
        confidence = DEFAULT_GEOMETRY_CONFIDENCE

    geometry = field(body, "geometry")
    associated = field(body, "associatedEntityId")
    end = field(body, "timePeriodEnd")
    label = field(body, "timePeriodLabel")

    entity_data: dict[str, Any] = {
        "name": field(body, "name"),
        "geometry": geometry,
        "geometrySerialized": serialize_geometry(geometry),
        "drawingMode": (
            "polygon"
            if isinstance(geometry, dict) and geometry.get("type") == "Polygon"
            else "polyline"
        ),
        "source": DRAWN_PROVENANCE,
        "associatedEntityId": associated,
        "timePeriodStart": field(body, "timePeriodStart"),
        "timePeriodEnd": None if end is MISSING else end,
    }
    # `timePeriodLabel: undefined` is dropped by `JSON.stringify`; an explicit
    # null the client sent is kept.
    if label is not MISSING:
        entity_data["timePeriodLabel"] = label

    description = field(body, "description")
    if truthy(description):
        entity_data["description"] = description
    if entity_type == "language-range":
        entity_data["languageId"] = associated

    sources = field(body, "sources")
    contribution: dict[str, Any] = {
        "entityType": entity_type,
        "action": "add",
        "entityId": associated,
        "entityData": entity_data,
        "sources": (
            sources if has_sources(sources) else [{"title": "User-drawn geometry"}]
        ),
        "confidence": confidence,
    }
    for key in ("notes", "contributorName", "contributorEmail"):
        value = field(body, key)
        if value is not MISSING:
            contribution[key] = value
    return contribution


def targets() -> dict[str, Any]:
    """The valid drawing targets, for the client's target selector."""
    return {"targets": list(DRAWN_GEOMETRY_TARGETS)}


__all__ = [
    "DEFAULT_GEOMETRY_CONFIDENCE",
    "DRAWN_GEOMETRY_TARGETS",
    "DRAWN_PROVENANCE",
    "LINE_TARGETS",
    "POLYGON_TARGETS",
    "ValidationResult",
    "drawn_geometry_to_contribution",
    "serialize_geometry",
    "targets",
    "validate_drawn_geometry",
    "validate_geometry",
]
